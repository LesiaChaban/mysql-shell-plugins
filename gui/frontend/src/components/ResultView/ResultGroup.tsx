/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

//import constructionIcon from "../../assets/images/under-construction.svg";

import { Component, IComponentProperties, IComponentState, ITabviewPage, TabPosition, Tabview } from "../ui";
import { ResultView } from ".";
import { IResultSet, IResultSetRows } from "../../script-execution";

export interface IResultGroupProperties extends IComponentProperties {
    resultSet: IResultSet;

    onResultPageChange?: (requestId: string, currentPage: number, sql: string) => void;
}

interface IResultGroupState extends IComponentState {
    selectedTab: string;
}

export class ResultGroup extends Component<IResultGroupProperties, IResultGroupState> {

    private resultRef = React.createRef<ResultView>();

    public constructor(props: IResultGroupProperties) {
        super(props);

        this.state = {
            selectedTab: "",
        };
    }

    public render(): React.ReactNode {
        const { resultSet, onResultPageChange } = this.props;
        const { selectedTab } = this.state;

        const pages: ITabviewPage[] = [];

        const resultView = <ResultView
            key="resultSet"
            ref={this.resultRef}
            resultSet={resultSet}
            onResultPageChange={onResultPageChange}
        />;

        pages.push({ id: "resultSet", caption: "RS", tooltip: "Result Set", content: resultView });

        // TODO: Temporarily disabled. Add a mode to the tabview where no tab is shown and tabs are switched
        // programmatically.
        /*pages.push({ id: "textResult", caption: "TX", tooltip: "Textual Output", content: icon });
        pages.push({ id: "formEditor", caption: "FE", tooltip: "Form Editor", content: icon });
        pages.push({ id: "fieldTypes", caption: "FT", tooltip: "Field Types", content: icon });
        pages.push({ id: "queryStats", caption: "QS", tooltip: "Query Statistics", content: icon });
        pages.push({ id: "executionPlan", caption: "EP", tooltip: "Execution Plan", content: icon });*/

        return (
            <Tabview
                className="resultGroup"
                stretchTabs={false}
                selectedId={selectedTab !== "" ? selectedTab : "resultSet"}
                tabPosition={TabPosition.Right}
                pages={pages}
                hideSingleTab

                onSelectTab={this.handleTabSelection}
            />
        );
    }

    public async addData(newData: IResultSetRows, replace: boolean): Promise<void> {
        if (this.resultRef.current) {
            await this.resultRef.current.addData(newData, replace);
        }
    }

    private handleTabSelection = (id: string): void => {
        this.setState({ selectedTab: id });
    };
}
