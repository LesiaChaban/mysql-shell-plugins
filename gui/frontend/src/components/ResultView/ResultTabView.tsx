/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

import { Component, IComponentProperties, IComponentState, ITabviewPage, TabPosition, Tabview } from "../ui";
import { IResultSet, IResultSetRows, IResultSets } from "../../script-execution";
import { ResultGroup } from ".";
import { IColumnInfo } from "../../app-logic/Types";

export interface IResultTabViewProperties extends IComponentProperties {
    resultSets: IResultSets;

    onResultPageChange?: (requestId: string, currentPage: number, sql: string) => void;
}

interface IResultTabViewState extends IComponentState {
    selectedTab: string;

    // React refs to the used ResultGroup instances, keyed by the request ID for the result set.
    groupRefs: Map<string, React.RefObject<ResultGroup>>;

    // Result IDs that are marked for purge.
    purgePending: Set<string>;
}

// Holds a collection of result views and other output in a tabbed interface.
export class ResultTabView extends Component<IResultTabViewProperties, IResultTabViewState> {

    public constructor(props: IResultTabViewProperties) {
        super(props);

        this.state = {
            selectedTab: "",
            groupRefs: new Map(),
            purgePending: new Set(),
        };
    }

    public render(): React.ReactNode {
        const { resultSets, onResultPageChange } = this.props;
        const { selectedTab, groupRefs } = this.state;

        const pages = resultSets.sets.map((resultSet: IResultSet, index: number): ITabviewPage => {
            const ref = React.createRef<ResultGroup>();
            groupRefs.set(resultSet.head.requestId, ref);

            return {
                id: `resultGroup${index}`,
                caption: `Result ${index + 1}`,
                content: (
                    <ResultGroup
                        ref={ref}
                        resultSet={resultSet}

                        onResultPageChange={onResultPageChange}
                    />
                ),
            };
        });

        return (
            <Tabview
                className="resultHost"
                stretchTabs={false}
                hideSingleTab={true}
                selectedId={selectedTab !== "" ? selectedTab : "resultGroup0"}
                tabPosition={TabPosition.Top}
                pages={pages}

                onSelectTab={this.handleTabSelection}
            />
        );
    }

    /**
     * Triggers a column update in the group which belongs to the given request id.
     *
     * @param requestId The actual request ID which identifies a specific result group.
     * @param columns The columns to set.
     */
    public async updateColumns(requestId: string, columns: IColumnInfo[]): Promise<void> {
        const { groupRefs } = this.state;

        const groupRef = groupRefs.get(requestId);
        if (groupRef && groupRef.current) {
            await groupRef.current.updateColumns(columns);
        }
    }

    /**
     * In order to update large result sets without re-rendering everything we use direct methods
     * to add new data.
     *
     * @param newData The data to add.
     *
     * @returns A promise the resolves when the operation is complete.
     */
    public async addData(newData: IResultSetRows): Promise<void> {
        const { groupRefs, purgePending } = this.state;

        const needPurge = purgePending.delete(newData.requestId);
        const groupRef = groupRefs.get(newData.requestId);
        if (groupRef && groupRef.current) {
            await groupRef.current.addData(newData, needPurge);
        }
    }

    /**
     * Moves existing data for the old request ID to a new ID and sends the underlying group a call to
     * note that next incoming data has to replace existing data.
     * We don't remove existing data here, to avoid flickering.
     *
     * @param oldRequestId The ID under which an existing group is registered.
     * @param newRequestId The ID under which this group is accessible after return.
     */
    public reassignData(oldRequestId: string, newRequestId: string): void {
        const { groupRefs, purgePending } = this.state;

        const groupRef = groupRefs.get(oldRequestId);
        if (groupRef && groupRef.current) {
            groupRefs.delete(oldRequestId);
            groupRefs.set(newRequestId, groupRef);

            purgePending.add(newRequestId);
        }
    }

    public markPendingReplace(requestId: string): void {
        const { purgePending } = this.state;

        purgePending.add(requestId);
    }

    private handleTabSelection = (id: string): void => {
        this.setState({ selectedTab: id });
    };

}
