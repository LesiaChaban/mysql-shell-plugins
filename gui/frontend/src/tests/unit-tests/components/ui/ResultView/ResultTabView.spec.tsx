/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { render } from "@testing-library/preact";
import { mount } from "enzyme";
import React from "react";
import { IResultTabViewProperties, ResultTabView } from "../../../../../components/ResultView";

describe("Result tabview tests", (): void => {

    it("Result tabview elements", () => {
        const component = mount<IResultTabViewProperties>(
            <ResultTabView
                resultSets={{
                    type: "resultSets",
                    sets: [{
                        head: {
                            sql: "select 1",
                            requestId: "123",
                        },
                        data: {
                            requestId: "123",
                            columns: [],
                            rows: [],
                            currentPage: 0,
                        },
                    }, {
                        head: {
                            sql: "select 2",
                            requestId: "456",
                        },
                        data: {
                            requestId: "456",
                            columns: [],
                            rows: [],
                            currentPage: 10,
                        },
                    }],
                }}
            />,
        );

        const props = component.props();
        expect(props.resultSets.sets.length).toEqual(2);

        component.unmount();
    });

    it("Result tabview (Snapshot) 1", () => {
        const component = render(
            <ResultTabView
                resultSets={{
                    type: "resultSets",
                    sets: [{
                        head: {
                            sql: "select 1",
                            requestId: "123",
                        },
                        data: {
                            requestId: "123",
                            columns: [],
                            rows: [],
                            currentPage: 0,
                        },
                    }, {
                        head: {
                            sql: "select 2",
                            requestId: "456",
                        },
                        data: {
                            requestId: "456",
                            columns: [],
                            rows: [],
                            currentPage: 10,
                        },
                    }],
                }}
            />,
        );
        expect(component).toMatchSnapshot();
    });

});
