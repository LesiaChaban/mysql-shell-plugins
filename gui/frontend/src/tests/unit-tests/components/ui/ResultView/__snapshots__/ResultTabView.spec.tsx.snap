// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result tabview tests Result tabview (Snapshot) 1 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    theme="dark"
  >
    <div>
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar resultHost"
        style="flex-direction: column;"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar tabview top resultTabview"
          style="flex-direction: column;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar selector horizontal tabArea"
            style="flex-direction: row;"
          >
            <div
              __self="[object Object]"
              __source="[object Object]"
              class="msg container fixedScrollbar selectorItem stepDown"
              id="selectorItemstepDown"
              style="flex-direction: row; justify-content: center; align-items: center; display: none;"
            />
            <div
              __self="[object Object]"
              __source="[object Object]"
              class="msg container fixedScrollbar selectorItem selected normal"
              id="123"
              style="flex-direction: row; justify-content: center; align-items: center;"
            >
              <label
                __self="[object Object]"
                __source="[object Object]"
                class="msg label"
                data-tooltip="inherit"
              >
                Result #0
              </label>
            </div>
            <div
              __self="[object Object]"
              __source="[object Object]"
              class="msg container fixedScrollbar selectorItem normal"
              id="456"
              style="flex-direction: row; justify-content: center; align-items: center;"
            >
              <label
                __self="[object Object]"
                __source="[object Object]"
                class="msg label"
                data-tooltip="inherit"
              >
                Result #0
              </label>
            </div>
            <div
              __self="[object Object]"
              __source="[object Object]"
              class="msg container fixedScrollbar selectorItem stepUp"
              id="selectorItemstepUp"
              style="flex-direction: row; justify-content: center; align-items: center; display: none;"
            />
          </div>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar content"
            style="flex-direction: column;"
          >
            <div
              __self="[object Object]"
              __source="[object Object]"
              class="msg container fixedScrollbar resultView"
              resultset="[object Object]"
              style="flex-direction: column;"
            >
              <div
                __self="[object Object]"
                __source="[object Object]"
                class="msg treeGrid horizontalGrid verticalGrid"
                style="font-size: 10pt;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="msg container fixedScrollbar resultHost"
      style="flex-direction: column;"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar tabview top resultTabview"
        style="flex-direction: column;"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar selector horizontal tabArea"
          style="flex-direction: row;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar selectorItem stepDown"
            id="selectorItemstepDown"
            style="flex-direction: row; justify-content: center; align-items: center; display: none;"
          />
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar selectorItem selected normal"
            id="123"
            style="flex-direction: row; justify-content: center; align-items: center;"
          >
            <label
              __self="[object Object]"
              __source="[object Object]"
              class="msg label"
              data-tooltip="inherit"
            >
              Result #0
            </label>
          </div>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar selectorItem normal"
            id="456"
            style="flex-direction: row; justify-content: center; align-items: center;"
          >
            <label
              __self="[object Object]"
              __source="[object Object]"
              class="msg label"
              data-tooltip="inherit"
            >
              Result #0
            </label>
          </div>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar selectorItem stepUp"
            id="selectorItemstepUp"
            style="flex-direction: row; justify-content: center; align-items: center; display: none;"
          />
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar content"
          style="flex-direction: column;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar resultView"
            resultset="[object Object]"
            style="flex-direction: column;"
          >
            <div
              __self="[object Object]"
              __source="[object Object]"
              class="msg treeGrid horizontalGrid verticalGrid"
              style="font-size: 10pt;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
